index.js

```diff
@@ -750,16 +1263,37 @@
                 return all(values, t.test) ? [t] : [];
               });
               if (isEmpty(okTypes)) {
-                throw invalidValue(name, map(types, nest), _value, index);
+                var prevType = head(types);
+                var prevValues = head(prevType.values);
+                throw conflictingTypeVar(
+                  signature(name, constraints, expArgTypes, expRetType),
+                  typeVarName,
+                  head(prevType.indexes),
+                  head(prevValues),
+                  valueTypes(prevValues),
+                  index,
+                  values[0],
+                  valueTypes(values)
+                );
               }
             } else {
-              okTypes = chain(commonTypes(map(values, determineActualTypes)),
-                              rejectInconsistent);
+              okTypes = valueTypes(values);
               if (isEmpty(okTypes) && !isEmpty(values)) {
                 throw orphanArgument(env, name, _value, index);
               }
             }
-            if (!isEmpty(okTypes)) $typeVarMap[typeVarName] = okTypes;
+            if (!isEmpty(okTypes)) {
+              if (typeVarName in $typeVarMap) {
+                map($typeVarMap[typeVarName],
+                    function(t) { t.indexes.push(index); });
+              } else {
+                $typeVarMap[typeVarName] = map(okTypes,
+                                               merge({
+                                                 indexes: [index],
+                                                 values: [values]
+                                               }));
```

Comment:
This indentation looks a bit strange to me. I'd prefer:

```javascript
$typeVarMap[typeVarName] =
map(okTypes, merge({indexes: [index], values: [values]}));
```

Indent the second line if you like. :)